website/content/en/highlights/2022-05-12-vector-config-subcommand

```diff
@@ -0,0 +1,227 @@
+---
+date: "2022-05-12"
+title: "Vector Config Subcommand"
+description: "A new subcommand to output configuration(s) as a standard JSON object"
+authors: ["001wwang"]
+pr_numbers: []
+release: "0.22.0"
+hide_on_release_notes: false
+---
+
+We've added a new subcommand `vector config` to help format and standardize your
+configurations. This can be useful when checking configurations into version
+control.
```

Comment:
Is version control a good use case for this? I'm struggling to identify why users would use this outside of computing a consistent hash. We don't even store this in our database, right?

Reply:
I think so. I imagine what's important to keep track of/view diffs of in version control is semantic differences rather than stylistic ones; this canonical format can thus be useful.

Correct, we currently don't store the output of this in our database, but that raises an interesting question. I can see us unifying what happens here and what happens in enterprise reporting (i.e. add this step as pre-processing). What do you think? cc: @leebenson

Reply:
For enterprise, it's important we store the original file since we want to allow human editing and management. Hence why I don't think it's a good idea to use this in version control. Most teams will probably want to manage and check in the original TOML file as opposed to the JSON output.

To be honest, I'm not sure why we even have this CLI comment. This seems like an internal implementation detail for computing our configuration hashes.

Reply:
That's fair. I got a bit turned around since we currently don't store the original file either (it gets translated to JSON, presumably to make its way through the soon-to-be-replaced JSON:API endpoint).

I think there might be some confusion here though. This command is unrelated to how we compute hashes (for now?) which we do with a separate `ConfigBuilderHash` struct here: https://github.com/vectordotdev/vector/blob/master/src/config/builder.rs#L288. 

I'll let @leebenson weigh in as I feel I'm missing some context in these areas.

Reply:
@blt a `vector fmt` command is a better approach for that, assuming it would format YAML and TOML. The direct JSON output designed for computing consistent hashes is not very human friendly, hence the reason configs are written in TOML/YAML.

My broader point is that I'm failing to see an actual use case here. Why would we encourage checking this in as opposed to the original TOML/YAML files? This creates 1) unnecessary overhead (like this PR) 2) unnecessary discussion (like what's happening now) 3) noise/confusion for the user.

Reply:
> @blt a `vector fmt` command is a better approach for that, assuming it would format YAML and TOML. The direct JSON output designed for computing consistent hashes is not very human friendly, hence the reason configs are written in TOML/YAML.

Ah, okay, I had misunderstood some of what was being done here. Yes, putting my user hat on, what I would want to store in VC would be the human readable toml/yaml. Where I might store the output of this command would be in CI -- maybe -- if we had a multi-stage config apply process, must like how `terraform plan` outputs a bunch of json that `terraform apply` then ingests. 

That is, would like a `vector fmt`, agree with Ben's point here, do like the property testing I'm seeing.

Reply:
(Copied from Slack, for posterity):

Yeah, it was primarily for our own testing. One slight utility for a public user is that it can turn a folder of configs into a single blob, which potentially makes passing it around easier. It’s basically a way to normalise a config.

If it’s unlikely to be helpful, we could:

- Nix the command (although there may still be utility in us testing configs later)
- Make the command part of a dev-only / feature flag build, so we can still use it if we want, but it isn’t exposed by default.
- Rename it to something more diagnostic-sounding (I like `vector fmt`).

`vector config` is being actively used for testing, so I'd like to keep it in some form.

Reply:
> Make the command part of a dev-only / feature flag build, so we can still use it if we want, but it isn’t exposed by default.

If there's utility for private use this approach works for me.

> Rename it to something more diagnostic-sounding (I like vector fmt).

FWIW if we were to make this `vector fmt` as-is I don't think it's suitable; a tool like that would want to keep the human readability aspect.

Reply:
I think a path forward (for now) can be:

- Remove the highlight article and public docs to avoid user confusion
- Keep the subcommand as it has proven useful for testing (cc: @sghall) but keep it hidden: https://docs.rs/clap/latest/clap/struct.App.html#method.hide. This allows frontend/others to continue using the normal `vector` binary without having to think about a dev-only build.
- Close out OP-336. Adjust #12527 which already documents outputting `TOML/YAML` as a future feature. At some point, we can rename/refactor this command into `vector fmt`. At the moment, this effectively seems like `vector fmt` for `JSON`.

How does that sound: @leebenson, @binarylogic ?

Reply:
Works for me

Reply:
@001wwang, that would be great 👍🏻